File: Platform/Protocol/Allors.Protocol/Compressor.cs
// <copyright file="Compressor.cs" company="Allors bvba">
// Copyright (c) Allors bvba. All rights reserved.
// Licensed under the LGPL license. See LICENSE file in the project root for full license information.
// </copyright>

namespace Allors.Protocol
{
    using System.Collections.Generic;

    public class Compressor
    {
        public const char IndexMarker = '~';

        public const char ItemSeparator = ',';

        private readonly Dictionary<string, string> keyByValue;

        private int counter;

        public Compressor()
        {
            this.keyByValue = new Dictionary<string, string>();
            this.counter = 0;
        }

        public string Write(string value)
        {
            if (value == null)
            {
                return null;
            }

            if (this.keyByValue.TryGetValue(value, out var key))
            {
                return key;
            }

            key = (++this.counter).ToString();
            this.keyByValue.Add(value, key);
            return $"{IndexMarker}{key}{IndexMarker}{value}";
        }
    }
}

File: Core/Workspace/Typescript.legacy/Site/wwwroot/Core/Workspace/Protocol/Compressor.ts
namespace Allors.Protocol {
  export class Compressor {
    static readonly indexMarker = '~';
    static readonly itemSeparator = ',';

    private keyByValue: { [k: string]: string };
    private counter: number;

    constructor() {
      this.keyByValue = {};
      this.counter = 0;
    }

    public write(value: string): string {
      if (value === undefined || value === null) {
        return null;
      }

      if (this.keyByValue.hasOwnProperty(value)) {
        return this.keyByValue[value];
      }

      const key = (++this.counter).toString();
      this.keyByValue[value] = key;
      return `${Compressor.indexMarker}${key}${Compressor.indexMarker}${value}`;
    }

  }
}
